dotnet/src/Connectors/Connectors.Memory.SqlServer/SqlServerClient.cs

```diff
@@ -0,0 +1,265 @@
+        await cmd.ExecuteNonQueryAsync(cancellationToken).ConfigureAwait(false);
+    }
+
+    /// <inheritdoc/>
+    public async Task DeleteBatchAsync(string tableName, IEnumerable<string> keys, CancellationToken cancellationToken = default)
+    {
+        var fullTableName = this.GetFullTableName(tableName);
+        var keysList = keys.ToList();
+        var keysParams = string.Join(", ", keysList.Select((_, i) => $"@k{i}"));
+        using var connection = new SqlConnection(this._connectionString);
+        await connection.OpenAsync(cancellationToken).ConfigureAwait(false);
+        using var cmd = connection.CreateCommand();
+        cmd.CommandText = $"""
+            DELETE
+            FROM {fullTableName}
+            WHERE [key] IN ({keysParams})
+            """;
+        for (var i = 0; i < keysList.Count; i++)
+        {
+            cmd.Parameters.AddWithValue($"k{i}", keysList[i]);
+        }
+        await cmd.ExecuteNonQueryAsync(cancellationToken).ConfigureAwait(false);
+    }
+
+    /// <inheritdoc/>
+    public async IAsyncEnumerable<(SqlServerMemoryEntry, double)> GetNearestMatchesAsync(string tableName, ReadOnlyMemory<float> embedding, int limit, double minRelevanceScore = 0, bool withEmbeddings = false, [EnumeratorCancellation] CancellationToken cancellationToken = default)
+    {
+        var queryColumns = withEmbeddings
+            ? "[key], [metadata], [timestamp], 1 - VECTOR_DISTANCE('cosine', [embedding], JSON_ARRAY_TO_VECTOR(@e)) AS [cosine_similarity], VECTOR_TO_JSON_ARRAY([embedding]) AS [embedding]"
+            : "[key], [metadata], [timestamp], 1 - VECTOR_DISTANCE('cosine', [embedding], JSON_ARRAY_TO_VECTOR(@e)) AS [cosine_similarity]";
+        var fullTableName = this.GetFullTableName(tableName);
+        using var connection = new SqlConnection(this._connectionString);
+        await connection.OpenAsync(cancellationToken).ConfigureAwait(false);
+        using var cmd = connection.CreateCommand();
+        cmd.CommandText = $"""
+            WITH data as (
+                SELECT {queryColumns}
+                FROM {fullTableName}
+            )
+            SELECT TOP (@limit) *
+            FROM data
+            WHERE [cosine_similarity] >= @score
+            ORDER BY [cosine_similarity] DESC
+            """;
+        cmd.Parameters.AddWithValue("@e", this.SerializeEmbedding(embedding));
+        cmd.Parameters.AddWithValue("@limit", limit);
+        cmd.Parameters.AddWithValue("@score", minRelevanceScore);
+        using var reader = await cmd.ExecuteReaderAsync(cancellationToken).ConfigureAwait(false);
+        while (await reader.ReadAsync(cancellationToken).ConfigureAwait(false))
+        {
+            var cosineSimilarity = reader.GetDouble(reader.GetOrdinal("cosine_similarity"));
+            yield return (this.ReadEntry(reader, withEmbeddings), cosineSimilarity);
+        }
+    }
+
+    private string GetFullTableName(string tableName) => $"{this._schema}.[{tableName}]";
```

Comment:
We may want to think about escaping here, i.e. for SQL injection... In other words, the user-provided `tableName` is probably not expected to be already escaped/sanitized (same for the `_schema`, which we can sanitize once when instantiating the class).

Probably good to do another pass to make sure there's nothing else that may be coming to the user that's getting interpolated into SQL (I don't think so, but better make sure).